src/riot-rs-threads/src/arch/riscv.rs

```diff
@@ -0,0 +1,135 @@
+
+    /// Triggers software interrupt for the context switch.
+    fn schedule() {
+        unsafe {
+            (&*SYSTEM::PTR)
+                .cpu_intr_from_cpu_3()
+                .modify(|_, w| w.cpu_intr_from_cpu_3().set_bit());
+        }
+    }
+
+    /// On RISC-V (ESP32), the stack doesn't need to be set up with any register values since
+    /// they are restored from the stored [`TrapFrame`].
+    fn setup_stack(thread: &mut Thread, stack: &mut [u8], func: usize, arg: usize) {
+        let stack_start = stack.as_ptr() as usize;
+        // 16 byte alignment.
+        let stack_pos = (stack_start + stack.len()) & 0xFFFFFFE0;
+        // Set up PC, SP, RA and first argument for function.
+        thread.sp = stack_pos;
+        thread.data.sp = stack_pos;
+        thread.data.a0 = arg;
+        thread.data.ra = cleanup as usize;
+        thread.data.pc = func;
+    }
+
+    /// Enable and trigger the appropriate software interrupt.
+    fn start_threading(_next_sp: usize) {
+        interrupt::disable(EspHalCpu::ProCpu, Interrupt::FROM_CPU_INTR3);
+        Self::schedule();
+        /// TODO: handle unwrap error?
+        interrupt::enable(Interrupt::FROM_CPU_INTR3, interrupt::Priority::min()).unwrap();
+    }
+}
+
+const fn default_trap_frame() -> TrapFrame {
+    TrapFrame {
+        ra: 0,
+        t0: 0,
+        t1: 0,
+        t2: 0,
+        t3: 0,
+        t4: 0,
+        t5: 0,
+        t6: 0,
+        a0: 0,
+        a1: 0,
+        a2: 0,
+        a3: 0,
+        a4: 0,
+        a5: 0,
+        a6: 0,
+        a7: 0,
+        s0: 0,
+        s1: 0,
+        s2: 0,
+        s3: 0,
+        s4: 0,
+        s5: 0,
+        s6: 0,
+        s7: 0,
+        s8: 0,
+        s9: 0,
+        s10: 0,
+        s11: 0,
+        gp: 0,
+        tp: 0,
+        sp: 0,
+        pc: 0,
+        mstatus: 0,
+        mcause: 0,
+        mtval: 0,
+    }
+}
+
+/// Copies the register state from `stc` to `dst`.
+///
+/// It copies state from the `TrapFrame` except for CSR registers
+/// `mstatus`, `mcuase` and `mtval`.
+fn copy_registers(src: &TrapFrame, dst: &mut TrapFrame) {
+    let (mstatus, mcause, mtval) = (dst.mstatus, dst.mcause, dst.mtval);
+    dst.clone_from(src);
+    dst.mstatus = mstatus;
+    dst.mcause = mcause;
+    dst.mtval = mtval;
+}
+
+/// Handler for software interrupt 3, which we use for context switching.
+#[allow(non_snake_case)]
+#[interrupt]
+fn FROM_CPU_INTR3(trap_frame: &mut TrapFrame) {
+    unsafe {
+        // clear FROM_CPU_INTR3
+        (&*SYSTEM::PTR)
+            .cpu_intr_from_cpu_3()
+            .modify(|_, w| w.cpu_intr_from_cpu_3().clear_bit());
+
+        sched(trap_frame)
+    }
+}
+
+unsafe fn sched(trap_frame: &mut TrapFrame) {
+    unsafe {
+        let cs = CriticalSection::new();
+        let next_pid = loop {
+            if let Some(pid) = (&*THREADS.as_ptr(cs)).runqueue.get_next() {
+                break pid;
+            }
+            riscv::asm::wfi();
```

Comment:
So with the interrupt priority of this ISR set to `min()` (at enable time, line 44), interrupts with higher priority preempt this one?

If no, this `wfi` will stop the CPU until an interrupt occurs, but not actually serve it (b/c this interrupt cannot be preempted).
If yes (which is the case IIUC), you'll actually need to wrap `sched()` in a critical section.
Take a look at the [arm assembly for PendSV](https://github.com/future-proof-iot/RIOT-rs/pull/182/files#diff-92b737e75a3c97cb0e16ed3a87c5649dee59536c7a9dc3a182a16540235315b3R158-R160), it wraps the call to `sched()` in `cpsid i` / `cpsie i`, and that's why it does the (hacky) enable/disable irq around `wfi`.

Reply:
(if `sched()` gets preempted by an isr, as soon as that ISR touches THREADS, it breaks the single &mut guarantee.)

Reply:
Isn't the critical section (line 116) sufficient here because the `riscv` `critical-section-single-hart` implementation disables all interrupts? I know the cortex-m implementation has both, the `cpsid`/`cpsie` calls and the critical section, but I didn't fully understand why.

Either way, I do need the enable/disable the interrupts around the `wfi`, good catch!

Reply:
Wdyt of 7036f86? 

It's a bit different than the cortex-m implementation because I am manually releasing & acquiring the `critical_section` around the WFI instead of enabling/ disabling any interrupts. 
This makes the code slightly more complex, but has the advantage that we don't make any assumptions about the `critical_section` implementation, because we don't actually know what interrupts they are enabling/ disabling.

Reply:
nice idea to use `acquire()` / `release()` directly. Didn't think of that when doing #187.

But here, doesn't `THREADS.with_mut()` acquire a critical section itself, causing a deadlock?

Reply:
> nice idea to use `acquire()` / `release()` directly. Didn't think of that when doing #187.

I actually prefer your version moving the critical section inside the loop. I stole that idea now :grin:: 2c47bcf.

> But here, doesn't `THREADS.with_mut()` acquire a critical section itself, causing a deadlock?

Nested critical sections are allowed (implemented as NOP) ([ref](https://docs.rs/critical-section/latest/critical_section/fn.with.html)). I only used it because the `EnsureOnce` API requires a cs.